index.bs

```diff
@@ -232,27 +232,137 @@ Non-Functional Requirements {#requirements-non-functional}
+from an advertising endpoint (acting as a QUIC server) or an advertising
+endpoint (acting as a QUIC server) wishes to send message to a discovering
+endpoint (acting as a client endpoint), it may wish to keep the QUIC connection
+alive.  Once neither side needs to keep the connection alive for the purposes of
+sending or receiving messages, the connection should be closed with an error
+code of 1.  In order to keep a QUIC connection alive, an endpoint may send an
+endpoint-status-request message, and any endpoint that receives an
+endpoint-status-request message should send an endpoint-status-response message.
+Such messages should be sent more frequently than the QUIC idle_timeout
+transport parameter (see section 18 of
+[QUIC](https://tools.ietf.org/html/draft-ietf-quic-transport-16)) and QUIC PING
+frames should not be used.  An idle_timeout transport parameter of 25 seconds is
+recommended.  If other mechanisms or traffic (such as other requests and
+responses) are serving the purpose of keeping the connection alive,
+endpoint-status-requests messages should be delayed or not used at all.
+
+If a client endpoint wishes to send messages to a server endpoint, the client
+endpoint can connect to the server endpoint "on demand"; it does not need to
+keep the connection alive.
+
+The endpoint-info-response message and endpoint-status-response
+messages may be extended to include additional information not defined
+in this spec.  If done ad hoc by applications and not in future specs,
+keys should be chosen to avoid collision, such as by choosing large
+integers or long strings.  Endpoints must ignore keys in the
+endpoint-info-message that it does not understand to allow endpoints
+to easily extend this message.
+
+Messages delivery using CBOR and QUIC streams {#control}
+========================================================
+
+Messages are serialized using [CBOR](https://tools.ietf.org/html/rfc7049).  To
+send a group of messages in order, that group of messages must be sent in one
+QUIC stream.  Different groups of messages (with no ordering dependency
+across groups) should be sent in different QUIC streams.  In order to put
+multiple CBOR-serialiezd messages into the the same QUIC stream, the following
+is used.
+
+For each message, the sender must write to the QUIC stream the following: 
+1. A type key representing the type of the message, encoded as a variable-length
+   integer (see Appendix A for type keys) 
+
+2. The message length encoded as a variable-length integer 
+
+3. The messge encoded as CBOR (whose length must match the prefixed length)
+
+If an endpoint receives a message for which it does not recognize a
+type key, it must close the QUIC connection with an application error
+code of 2 and should include the unknown type key in the reason phrase
+(see [QUIC](https://tools.ietf.org/html/draft-ietf-quic-transport-16)
+sections 19.4).
+
+Variable-length integers are encoded in the same format as defined by [QUIC
```

Comment:
Instead of a CBOR varint?  Since the implementation has to parse CBOR anyway, it might be easier to use CBOR here too.

Reply:
I think the CBOR varints are too big.  I went through many different designs and options for this, and this was the best mix of small and easy to implement.  If you read the spec I linked to, you'll see it's very easy to implement.  And if you work through the CBOR varints, I think you'll find that it will use up an extra byte per int (2 bytes per message) most of the time.

Reply:
SGTM.